src/gp_instance.cpp

```diff
@@ -174,3 +175,63 @@ void GPInstance::EstimateSBNParameters() {
   double marginal_log_lik = GetEngine()->GetLogMarginalLikelihood();
   std::cout << std::setprecision(9) << marginal_log_lik << std::endl;
 }
+
+RootedTreeCollection GPInstance::GetRootedTreeCollection() {
+
+  Tree::TreeVector tree_vector;
+  Node::NodePtrVec trees = dag_.GenerateAllTrees();
+  EigenVectorXd bl = engine_->GetBranchLengths();
+
+  // Leaves encoding to parent subsplit encoding.
+  std::unordered_map<const Node*, Bitset> leaves2subsplit_indexer;
+  for (size_t i = 0; i < trees.size(); i++) {
+    trees.at(i)->PreOrder([this, &leaves2subsplit_indexer](const Node *node)
+    {
+      GPDAGNode *dag_node = dag_.GetDagNode(node->Id());
+      if (!leaves2subsplit_indexer.count(node)) {
+        SafeInsert(leaves2subsplit_indexer, node, dag_node->GetBitset());
+      }
+
+      for (size_t i = 0; i < node->Children().size(); i++) {
+        size_t child_id = node->Children().at(i)->Id();
+        GPDAGNode *child_node = dag_.GetDagNode(child_id);
+        size_t i0 = dag_.GetGPCSPIndex(dag_node->GetBitset(false) + child_node->GetBitset());
+        size_t i1 = dag_.GetGPCSPIndex(dag_node->GetBitset(true) + child_node->GetBitset());
+        size_t gpcsp_idx = std::min(i0, i1);
+        Assert(gpcsp_idx < SIZE_MAX, "GPCSP does not exist.");
+      }
+    });
+  }
+
+  for (auto &root_node : trees) {
+    // Polish will re-assign the node Ids.
+    root_node->Polish();
```

Comment:
As suggested above, I suggest polishing as part of the topology construction process.

Reply:
If you prefer that Polish be called right after the topology is constructed, I will have to restructure the code.  I don't have preference either way, but if you have strong preference that let me know.

Recall that `GPDAG` does not have access to the branch lengths, which are in the `GPEngine`. So the way it works now is for `GPInstance` to 
1. Get the topologies from `GPDAG::GenerateAllTrees`, The ID assigned to each Node instance upon construction is the corresponding `GPDAGNode` ID. Note: calling Polish re-assigns node IDs.
2. Get the branch lengths from `GPEngine`,
3. Traverse each topology to visit each node, use their Node IDs to get the `GPDAGNode` instance and the corresponding `Bitset` representation, store them in an indexer (`NodePtr` to `Bitset`).
4. Do a second traversal, this time look up the index for PCSP and set the branch length.